sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -83,9 +83,18 @@ class StringType private (val collationId: Int) extends AtomicType with Serializ
   override def jsonValue: JValue = JString("string")
 
   override def equals(obj: Any): Boolean =
```

Comment:
it will be very tricky if we overwrite the low-level `equals` and `hashCode` to treat the default collation the same as utf8. We should leave the Java object equality as it is, and only treat then the same when we only care about semantic.

Reply:
But there were places in the code that compare `collationId` directly. As example, `CollationTypeCoercion` will not work properly if we do not overwrite `hashCode` method. Also there are other places in the code that are modified and directly compare `collationId` in this case which would behave wrong in this case. I would keep the `equals` method overriden just as it is in the PR.

Reply:
@cloud-fan But this is a user facing change, wouldn't it be counterintuitive if `StringType == StringType("UTF8_BINARY")` returned false, implying that default collation is not utf8 binary? @srielau do you have any opinions here?

Reply:
+1 with stefan on this, doesn't sound super intuitive to me

Reply:
This is an anti-pattern, we should never do it. Java object equality should compare what's in the object. For example, we do not override `LogicalPlan#equals` but add `LogicalPlan#semanticEquals`.

We should change the places that compare collation id directly, to probably call a util function that takes care of the default collation.

Reply:
In my opinion, although it may be the correct approach regarding equality of objects in Java/Scala, I think this can bring up more problems in the future even though I refactored places in the code where we compare the `collationId`. There are a lot other places that compare `DataType` directly and currently for `StringType`, this is being done by comparing the `collationId`. These places are very hard to detect, and I agree with Stefan that `StringType == StringType("UTF8_BINARY")` returning false here would be misleading. Furthermore, for all the code that is being added afterwards, people will have to be really careful when comparing `DataType`s, since the `equals`method does not work as expected for `StringType`.

Reply:
I get the both arguments but I thinks it's pretty evident that they have major downsides. @cloud-fan I don't really see how is any of this better than just annotating explicitly collated strings so that we can tell the difference. From my POV that approach shouldn't have any side-effect issues

Reply:
StirngType with un-determined collation is and should be different from StringType with udf8 collation. It's very dangerous to treat them the same at such a low level.

Let's list the places that should treat them the same:
1. Spark internal type coercion rules
2. Spark internal collation-aware functions/operators
3. external code that needs to deal with string type (UDF, data source)

For 1 and 2, I think most code goes to `CollationFactory` and we don't need to update a lot of places. For 3, there won't be any problem by default as collation is not enabled. The existing code still works as all `StringType` has un-determined collation. If users turn on collation, the existing code is broken anyway as they need to handle different collations.

Eventually, when collation is enabled, we should not have un-determined collation and all the string types should have a collation determined by the context (session default, or table/schema/catalog default).

Reply:
I agree that there are downsides of having `STRING COLLATE UTF8_BINARY` equals to `STRING`, however, I still think that even if we track and fix all the places where we should consider them the same, the change will be prone to errors when new code is being added by people not familiar with this case. Maybe the change that @stefankandic proposed is the best solution here.